docs/dotnet/dotnet-aspnet-metrics.md

```diff
@@ -0,0 +1,216 @@
+# Semantic Conventions for ASP.NEt Core specific metrics
+
+**Status**: [Experimental][DocumentStatus]
+
+This document defines semantic conventions for ASP.NET Core metrics, not specific to HTTP
+
+**Disclaimer:** These are initial .NET metric instruments available in .NET 8 but more may be added in the future.
+
+<!-- toc -->
+
+- [Routing](#routing)
+  * [Metric: `aspnet.routing.successful_matches`](#metric-aspnetroutingsuccessful_matches)
+  * [Metric: `aspnet.routing.failed_matches`](#metric-aspnetroutingfailed_matches)
+- [Exception metrics](#exception-metrics)
+  * [Metric: `aspnet.diagnostics_handler.exceptions`](#metric-aspnetdiagnostics_handlerexceptions)
+- [Rate-limiting](#rate-limiting)
+  * [Metric: `aspnet.rate_limiting.active_request_leases`](#metric-aspnetrate_limitingactive_request_leases)
+  * [Metric: `aspnet.rate_limiting.request_lease.duration`](#metric-aspnetrate_limitingrequest_leaseduration)
+  * [Metric: `aspnet.rate_limiting.queued_requests`](#metric-aspnetrate_limitingqueued_requests)
+  * [Metric: `aspnet.rate_limiting.queued_requests.duration`](#metric-aspnetrate_limitingqueued_requestsduration)
+  * [Metric: `aspnet.rate_limiting.rejected_requests`](#metric-aspnetrate_limitingrejected_requests)
+
+<!-- tocstop -->
+
+## Routing
+
+All routing metrics are reported by `Microsoft.AspNetCore.Routing` meter.
+
+### Metric: `aspnet.routing.successful_matches`
+
+**Status**: [Experimental][DocumentStatus]
+
+**TODO: can we try one aspnet.routing.matches** ?
+
+This metric is required.
+
+<!-- semconv metric.aspnet.routing.successful_matches(metric_table) -->
+| Name     | Instrument Type | Unit (UCUM) | Description    |
+| -------- | --------------- | ----------- | -------------- |
+| `aspnet.routing.successful_matches` | Counter | `{match}` | Number of requests that successfully matched to an endpoint. [1] |
+
+**[1]:** Meter name is `Microsoft.AspNetCore.Routing`.
+<!-- endsemconv -->
+
+<!-- semconv metric.aspnet.routing.successful_matches(full) -->
+| Attribute  | Type | Description  | Examples  | Requirement Level |
+|---|---|---|---|---|
+| `aspnet.routing.fallback` | string | TODO | `TODO` | Required |
+| `http.route` | string | The matched route (path template in the format used by the respective server framework). See note below [1] | `/users/:userID?`; `{controller}/{action}/{id?}` | Required |
+
+**[1]:** MUST NOT be populated when this is not supported by the HTTP server framework as the route attribute should have low-cardinality and the URI path can NOT substitute it.
+SHOULD include the [application root](/docs/http/http-spans.md#http-server-definitions) if there is one.
+<!-- endsemconv -->
+
+### Metric: `aspnet.routing.failed_matches`
+
+**TODO: would if be useful to add method name, scheme, or other attributes?**
```

Comment:
I'm not sure how these other attributes would be of much help so I'd leave it as-is unless there is a more concrete scenario in mind. I'm expecting developers would notice the counter is higher than expected, then switch over to logs to review what the unmatched input paths were.

Reply:
> I'm expecting developers would notice the counter is higher than expected, then switch over to logs to review what the unmatched input paths were.

Switching to logs to find the top unmatched reasons is hard, let's try to minimize this pain and reduce cognitive load.

The method can help narrow the problem down without going into the logs. Also, it won't be possible to add new attributes later, so let's add what other metrics have for the sake of consistency (scheme, method, server address/port)

Reply:
I saw what you added in #2 and its probably fine as long as the information is readily available. I still think if you care about diagnosing and resolving why routes aren't matching you will need to know what the inputs are that are trying to be matched which is going to require a log. (Putting that data in a metric attribute opens the door to unbounded cardinality that could be exploited by someone trying to DDOS a service).